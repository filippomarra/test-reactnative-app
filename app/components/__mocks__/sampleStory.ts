import Story from '../../models';

export const sampleStory: Story = {
    kind: 'story',
    id: 151723133,
    created_at: new Date('2017-10-05T15:30:45Z'),
    updated_at: new Date('2020-09-09T08:08:04Z'),
    accepted_at: new Date('2017-10-18T14:04:39Z'),
    story_type: 'chore',
    story_priority: 'p3',
    name: 'API Design',
    description:
        'As an engineer, I will produce a RAML file to discuss with the team the entry points and the payloads for the proxy so that we can start thinking how to mock this data in the frontend',
    current_state: 'accepted',
    requested_by_id: 1945893,
    url: 'https://www.pivotaltracker.com/story/show/151723133',
    project_id: 2116794,
    owner_ids: [2958113],
    labels: [
        {
            id: 19223788,
            project_id: 2116794,
            kind: 'label',
            name: 'debito-tecnico',
            created_at: new Date('2017-10-05T15:30:45Z'),
            updated_at: new Date('2020-09-09T08:08:04Z'),
        },
    ],
    owned_by_id: 2958113,
};
