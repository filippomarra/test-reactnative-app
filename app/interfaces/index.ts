import Story from '../models';

export interface ApiProps {
    setLoading: (loading: boolean) => void;
    setStories: (stories: Story[]) => void;
}

export interface AppHeaderProps {
    stories: Story[];
    setStories: (stories: Story[]) => void;
}

export interface ModalProps {
    visible: boolean;
    hideDialog: () => void;
    story: Story;
}

export interface StoryItemProps {
    index: number;
    showDialog: (story: Story) => void;
    story: Story;
}

export interface StoryScrollViewProps {
    isLoading: boolean;
    showDialog: (story: Story) => void;
    stories: Story[];
}
