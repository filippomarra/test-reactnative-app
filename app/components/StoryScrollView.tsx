import React from 'react';
import { ActivityIndicator, ScrollView, View } from 'react-native';

import { StoryScrollViewProps } from '../interfaces';
import { styles } from '../styles';

import { StoryItem } from './StoryItem';

export const StoryScrollView = (props: StoryScrollViewProps) => {
    const { isLoading, showDialog, stories } = props;

    return (
        <ScrollView>
            {isLoading ? (
                <ActivityIndicator
                    size="large"
                    style={styles.activityIndicator}
                    testID="loader"
                />
            ) : (
                <View testID="storyView">
                    {stories.map((item, index) => (
                        <StoryItem
                            story={item}
                            showDialog={showDialog}
                            index={index}
                        />
                    ))}
                </View>
            )}
        </ScrollView>
    );
};
