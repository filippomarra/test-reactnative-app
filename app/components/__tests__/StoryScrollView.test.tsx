import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';

import { StoryScrollViewProps } from '../../interfaces';

import { sampleStory } from '../__mocks__/sampleStory';
import { StoryScrollView } from '../StoryScrollView';

const defaultProps: StoryScrollViewProps = {
    isLoading: true,
    showDialog: jest.fn(),
    stories: [],
};

describe('StoryScrollView component tests', () => {
    it('renders correctly with default props', () => {
        expect(
            render(<StoryScrollView {...defaultProps} />).toJSON(),
        ).toMatchSnapshot();
    });

    describe('When isLoading is set to true', () => {
        it('should render the ActivityIndicator', () => {
            const { queryByTestId } = render(
                <StoryScrollView {...defaultProps} />,
            );
            expect(queryByTestId('loader')).not.toBeNull();
        });

        it('should not render the View', () => {
            const { queryByTestId } = render(
                <StoryScrollView {...defaultProps} />,
            );
            expect(queryByTestId('storyView')).toBeNull();
        });
    });

    describe('When isLoading is set to false', () => {
        it('should render the View', () => {
            const { queryByTestId } = render(
                <StoryScrollView {...defaultProps} isLoading={false} />,
            );
            expect(queryByTestId('storyView')).not.toBeNull();
        });

        it('should not render the ActivityIndicator', () => {
            const { queryByTestId } = render(
                <StoryScrollView {...defaultProps} isLoading={false} />,
            );
            expect(queryByTestId('loader')).toBeNull();
        });

        describe('When stories is empty', () => {
            it('should not render any StoryItem', () => {
                const { queryByTestId } = render(
                    <StoryScrollView {...defaultProps} isLoading={false} />,
                );
                const storyView = queryByTestId('storyView');
                expect(storyView?.children.length).toBe(0);
            });
        });

        describe('When stories is not empty', () => {
            it('should render the same number of StoryItem as the stories', () => {
                const { queryByTestId } = render(
                    <StoryScrollView
                        {...defaultProps}
                        isLoading={false}
                        stories={[sampleStory]}
                    />,
                );
                const storyView = queryByTestId('storyView');
                expect(storyView?.children.length).toBe(1);
            });
        });
    });
});
