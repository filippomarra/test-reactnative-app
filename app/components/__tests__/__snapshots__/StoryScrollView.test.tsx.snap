// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryScrollView component tests renders correctly with default props 1`] = `
<RCTScrollView>
  <View>
    <ActivityIndicator
      size="large"
      style={
        Object {
          "paddingTop": 15,
        }
      }
      testID="loader"
    />
  </View>
</RCTScrollView>
`;
